File: ontrack-test-utils/src/main/java/net/nemerosa/ontrack/test/TestUtils.java
package net.nemerosa.ontrack.test;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.TreeNode;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import net.nemerosa.ontrack.json.ObjectMapperFactory;
import org.apache.commons.io.IOUtils;
import org.apache.commons.lang3.Validate;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.text.SimpleDateFormat;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.concurrent.atomic.AtomicLong;

import static org.junit.Assert.assertEquals;

public final class TestUtils {

    private static final ObjectMapper mapper = ObjectMapperFactory.create();
    private static final AtomicLong counter = new AtomicLong();

    private TestUtils() {
    }

    public static List<Integer> range(int from, int to) {
        Validate.isTrue(to >= from, "'to' value must be greater or equal to the 'from' value.");
        List<Integer> l = new ArrayList<>();
        for (int i = from; i <= to; i++) {
            l.add(i);
        }
        return l;
    }

    public static String uid(String prefix) {
        return prefix + new SimpleDateFormat("mmssSSS").format(new Date()) + counter.incrementAndGet();
    }

    public static void assertJsonWrite(JsonNode expectedJson, Object objectToWrite) {
        assertEquals(
                expectedJson,
                mapper.valueToTree(objectToWrite)
        );
    }

    public static void assertJsonWrite(JsonNode expectedJson, Object objectToWrite, Class<?> viewClass) throws JsonProcessingException {
        assertJsonWrite(
                mapper,
                expectedJson,
                objectToWrite,
                viewClass
        );
    }

    public static void assertJsonWrite(ObjectMapper mapper, JsonNode expectedJson, Object objectToWrite, Class<?> viewClass) throws JsonProcessingException {
        assertEquals(
                mapper.writeValueAsString(expectedJson),
                mapper.writerWithView(viewClass).writeValueAsString(objectToWrite)
        );
    }

    public static <T> void assertJsonRead(T expectedResult, JsonNode jsonToRead, Class<T> type) throws JsonProcessingException {
        assertEquals(
                expectedResult,
                mapper.treeToValue(jsonToRead, type)
        );
    }

    public static void assertJsonEquals(TreeNode o1, TreeNode o2) {
        assertEquals(
                o1,
                o2
        );
    }

    public static LocalDateTime dateTime() {
        return LocalDateTime.of(2014, 5, 13, 20, 11);
    }

    public static String dateTimeJson() {
        return "2014-05-13T20:11:00Z";
    }

    public static @NotNull
    byte[] resourceBytes(@NotNull String path) throws IOException {
        return IOUtils.toByteArray(TestUtils.class.getResource(path));
    }

    public static JsonNode resourceJson(String path) throws IOException {
        return mapper.readTree(resourceBytes(path));
    }

}

File: ontrack-kdsl-acceptance/src/test/java/net/nemerosa/ontrack/kdsl/acceptance/tests/support/TestUtils.kt
package net.nemerosa.ontrack.kdsl.acceptance.tests.support

import com.fasterxml.jackson.databind.JsonNode
import kotlinx.coroutines.TimeoutCancellationException
import kotlinx.coroutines.delay
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.withTimeout
import net.nemerosa.ontrack.json.parseAsJson
import org.apache.commons.io.IOUtils
import org.junit.jupiter.api.fail
import java.nio.charset.Charset
import java.text.SimpleDateFormat
import java.time.Duration
import java.util.*
import java.util.concurrent.atomic.AtomicLong

/**
 * Sequence for unique IDs.
 */
private val counter = AtomicLong()

/**
 * Generates a unique ID.
 */
fun uid(prefix: String): String =
    prefix + SimpleDateFormat("mmssSSS").format(Date()) + counter.incrementAndGet()

/**
 * Waiting until a condition is met.
 */
fun waitUntil(
    initial: Long = 0L,
    timeout: Long = 60_000L,
    interval: Long = 10_000L,
    task: String? = null,
    onTimeout: (() -> Unit)? = null,
    condition: () -> Boolean,
) {
    runBlocking {
        if (initial > 0) {
            delay(initial)
        }
        try {
            withTimeout(timeout) {
                while (!condition()) {
                    delay(interval)
                }
            }
        } catch (ex: TimeoutCancellationException) {
            if (onTimeout != null) {
                onTimeout()
            }
            if (task != null) {
                fail("Timeout waiting for $task to complete.")
            } else {
                val message = """
                    Timeout waiting for a task to complete.
                    
                    No task name was provided, proving difficult to assert what exactly has timed out.
                    
                    Consider adding the `task` parameter to the call of the `waitUntil` method.
                """.trimIndent()
                fail(message, ex)
            }
        }
    }
}

/**
 * Seconds to millis
 */
val Int.seconds get() = Duration.ofSeconds(toLong()).toMillis()

/**
 * Reads a resource from the classpath as JSON
 */
fun resourceAsJson(path: String): JsonNode =
    resourceAsText(path).parseAsJson()

/**
 * Reads a resource from the classpath as text
 */
fun resourceAsText(path: String, charset: Charset = Charsets.UTF_8): String =
    resourceAsBytes(path).toString(charset)


/**
 * Reads a resource from the classpath as binary
 */
fun resourceAsBytes(path: String): ByteArray = IOUtils.toByteArray(TestUtils::class.java.getResource(path))

/**
 * Anchor class for the resource paths
 */
private class TestUtils


/**
 * Gets a value from the system properties or from the environment.
 */
fun getEnv(property: String): String {
    return getOptionalEnv(property)
        ?: throw IllegalStateException("Cannot find $property system property or ${propertyToEnvName(property)} environment variable.")
}

/**
 * Gets a value from the system properties or from the environment.
 */
fun getOptionalEnv(property: String): String? {
    val sysValue = System.getProperty(property)
    return if (sysValue.isNullOrBlank()) {
        val envName = propertyToEnvName(property)
        val envValue = System.getenv(envName)
        if (envValue.isNullOrBlank()) {
            null
        } else {
            envValue
        }
    } else {
        sysValue
    }
}

private fun propertyToEnvName(property: String) = property.uppercase().replace('.', '_')
